### PR TITLE
[WIP] plugin: envoy.grpc_http1_reverse_bridge

### PR DESCRIPTION
**Pull request by sp-manuel-jurado**
**[WIP] plugin: envoy.grpc_http1_reverse_bridge**

The objective of this PR is to implement a plugin to support the envoy "gRPC HTTP / 1.1 Reverse Bridge" filter.

Further info:
- [envoy http filter doc](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/grpc_http1_reverse_bridge_filter)
- [envoy api doc](https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/filter/http/grpc_http1_reverse_bridge/v2alpha1/config.proto)
- [gloo writing upstream plugins doc](https://docs.solo.io/gloo/latest/guides/dev/writing-upstream-plugins/)

Envoy config example:
```
...
    - filters:
      - name: envoy.http_connection_manager
        config:
          stat_prefix: ingress_http
          route_config:
            name: local_route
            virtual_hosts:
            - name: local_service
              domains: ["*"]
              routes:
              - match:
                  prefix: "/sp.rpc.PingService/"
                route:
                  cluster: ping
                per_filter_config:
                  envoy.filters.http.grpc_http1_reverse_bridge:
                    disabled: true
              - match:
                  prefix: "/"
                route:
                  cluster: backend
          http_filters:
            - name: envoy.filters.http.grpc_http1_reverse_bridge
              config:
                content_type: application/grpc+proto
                withhold_grpc_frames: true
...
```

### CODE REVIEW

**Comment by christian-posta** [comment]
@kdorosh  @yuval-k or @EItanya can provide comment

**Comment by yuval-k** [comment]
Hi @christian-posta , this is still a WIP (there are still unimplemented panics in the code). API looks ok to me

File: projects/gloo/api/v1/options.proto [hunk 1]
```diff
@@ -103,6 +104,11 @@ message HttpListenerOptions {
 
     // Enterprise-only: Proxy latency
     envoy.config.filter.http.proxylatency.v2.ProxyLatency proxy_latency = 9;
+
+    // Grpchttp1reversebridge is an HTTP option which enables envoy gRPC HTTP/1.1 reverse bridge HTTP filter
+    // This is a filter that enables converting an incoming gRPC request into a HTTP/1.1 request to allow a server
+    // that does not understand HTTP/2 or gRPC semantics to handle the request.
+    grpchttp1reversebridge.options.gloo.solo.io.Grpchttp1reversebridge grpchttp1reversebridge = 10;
```
Comment (kdorosh) [inline projects/gloo/api/v1/options.proto:111]: nit, let's rename the new config to `grpc_http1_reverse_bridge`

File: projects/gloo/api/v1/options/grpchttp1reversebridge/grpchttp1reversebridge.proto [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+message Config {
+    // The content-type to pass to the upstream when the gRPC bridge filter is applied.
+    // The filter will also validate that the upstream responds with the same content type.
+    string content_type = 1;
+
+    // If true, Envoy will assume that the upstream doesn’t understand gRPC frames and strip
+    // the gRPC frame from the request, and add it back in to the response. This will hide the
+    // gRPC semantics from the upstream, allowing it to receive and respond with a simple binary
+    // encoded protobuf.
+    string withhold_grpc_frames = 2;
```
Comment (kdorosh) [inline projects/gloo/api/v1/options/grpchttp1reversebridge/grpchttp1reversebridge.proto:15]: this sounds like a boolean value. if so, we shouldn't use a string here

Comment (kdorosh) [reply projects/gloo/api/v1/options/grpchttp1reversebridge/grpchttp1reversebridge.proto:15]: confirmed that the underlying envoy config is a boolean, thus this should be as well

File: projects/gloo/api/v1/options/grpchttp1reversebridge/grpchttp1reversebridge.proto [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+// Grpchttp1reversebridge per filter configuration
+// It used to allow disabling filter by route
+message PerFilterConfig {
+    // It disables filter
+    bool disabled = 1;
+}
```
Comment (kdorosh) [inline projects/gloo/api/v1/options/grpchttp1reversebridge/grpchttp1reversebridge.proto:23]: this isn't used in the proto anywhere. the plugin options in gloo are already nested under `PerFilterConfig`, thus the `disabled` entry should live on `Config` in `projects/gloo/api/v1/options/grpchttp1reversebridge/grpchttp1reversebridge.proto`

**Comment by yuval-k** [comment]
that's fine as i believe we can mix v2 and v3

**Pull request closed without merging**